dbms/src/Interpreters/Join.cpp

```diff
@@ -1123,17 +1199,30 @@ void NO_INLINE joinBlockImplTypeCase(
         {
             auto key_holder = key_getter.getKeyHolder(i, &pool, sort_key_containers);
             auto key = keyHolderGetKey(key_holder);
-            size_t segment_index = 0;
             size_t hash_value = 0;
-            if (map.getSegmentSize() > 0 && !ZeroTraits::check(key))
+            size_t segment_index = 0;
+            if (enable_fine_grained_shuffle)
+            {
+                /// In FineGrainedShuffleWriter, selector value finally maps to packet_stream_id
+                auto packet_stream_id = selector[i] % fine_grained_shuffle_count;
+                /// Relies on the facts that:
+                /// In ExchangeReceiver, build_stream_id = packet_stream_id % build_stream_count;
+                /// In Join, build_concurrency decides map's segment size, and build_steam_id decides the segment index
+                segment_index = packet_stream_id % segment_size;
+            }
+            else
```

Comment:
Does this really improve performance?
hash_value will still be calculated when map.find?

Reply:
prob won't benefit from fine grained shuffle, in fact it will cost a little due to the extra hash calculation. Build will benefit a lot, prob just add this logic to make it correct.

Reply:
So this is to ensure that build and probe side select the same segment_index?

Reply:
Yes, because prob side's exchange operator won't be set FinedGrainedShuffle info in TiDB, need to do this to check correctness.

Reply:
ok, we can add some comments about the reason why we do the extra hash calculation

Reply:
Done.